### DISCUSSION TITLE
How can I recreate a swarm with a different `PeerId`?

### DISCUSSION THREAD

**Discussion by kewbish** [libp2p/rust-libp2p#5731] [Q&A]
Is there a way to gracefully close a swarm and recreate it with a provided `PeerId`? Starting from the [chat example](https://github.com/libp2p/rust-libp2p/blob/master/examples/chat/src/main.rs), this is what I understand I'd need to do:
- Get all the listeners via `swarm.listeners()`
- Remove all the listeners via `swarm.remove_listener()`
- Rebuild the swarm with a new existing identity
- Resubscribe to Gossipsub topics and listen on the previous multiaddr

I'm curious if there's any cleaner way to do this, though.

My use case is that a peer may need to send/receive a Gossipsub message that informs it what its keypair/`PeerId` should be (during a login, for example), and the peer will need to switch to using its existing keypair and `PeerId` afterwards so the other nodes have some stable way of identifying accounts. Let me know if this model of how to use `PeerId`s is incorrect or could be improved.

Thanks in advance for any help!

**Comment by elenaf9** [comment]
> a peer may need to send/receive a Gossipsub message that informs it what its keypair/`PeerId` should be

You might also want to look into the [request-response](https://docs.rs/libp2p/latest/libp2p/request_response/index.html) protocol for direct 1:1 messages between two nodes. 

> - Get all the listeners via swarm.listeners()

> - Resubscribe to Gossipsub topics and listen on the previous multiaddr

In the chat example, we use `/ip4/0.0.0.0/` to listen on all interfaces, and `/0/quic-v1` and `/tcp/0` to listen on random ports. You could, instead of listening on a random port, just listen on a fixed port. That way the next time you start listening you can simply listen on the same port again. 
Note however that generally even after removing the listener, the OS might not directly allow you to listen again on the same port.

**Reply by kewbish** [reply]
Thanks, will give the `request-response` protocol a try!
